### PR TITLE
caf: Add click_detector

### PR DESCRIPTION
**Pull request by MarekPieta**
Change adds click detector module to CAF.

Jira: NCSDK-8080

### CODE REVIEW

**Review by pdunaj — APPROVED** [review]

File: subsys/caf/modules/Kconfig.click_detector [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor
```
Comment (emob-nordic) [inline subsys/caf/modules/Kconfig.click_detector:2]: ASA

File: subsys/caf/modules/Kconfig.click_detector [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	  Location of configuration file for click detector module.
+
+config CAF_CLICK_DETECTOR_PM_EVENTS
+	bool "Power management events support"
+	depends on CAF_PM_EVENTS
+	help
+	  React on power management events in click detector module.
+
+module = CAF_CLICK_DETECTOR
+module-str = click detector module
```
Comment (emob-nordic) [inline subsys/caf/modules/Kconfig.click_detector:27]:
``` suggestion
module-str = caf module click detector
```

File: include/caf/click_detector.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+* :c:enumerator:`CLICK_SHORT` - Button pressed and released after short time.
+* :c:enumerator:`CLICK_NONE` - Button pressed and held for a period of time that is too long for :c:enumerator:`CLICK_SHORT`, but too short for :c:enumerator:`CLICK_LONG`.
+* :c:enumerator:`CLICK_LONG` - Button pressed and held for a long period of time.
+* :c:enumerator:`CLICK_DOUBLE` - Two sequences of the button press and release in a short time interval.
+
+The exact values of time intervals for click types are defined in the :file:`subsys/caf/modules/click_detector.c` file.
+
+Configuration
+*************
+
```
Comment (emob-nordic) [inline include/caf/click_detector.rst:26]: Configuration was intended to start with list of options required for making the module work. See [buttons](https://github.com/nrfconnect/sdk-nrf/blob/master/include/caf/buttons.rst) for reference.

File: include/caf/click_detector.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	static const struct click_detector_config click_detector_config[] = {
+		{
+			.key_id = KEY_ID(0x00, 0x01),
+			.consume_button_event = false,
+		},
+	};
+
+.. note::
+   The :c:macro:`KEY_ID` macro is defined in :file:`include/caf/key_ids.h`.
```
Comment (emob-nordic) [inline include/caf/click_detector.rst:50]: Why don't you add `#include <caf/key_id.h>` in this code snippet? Btw: `s/key_ids/key_id`

**Review by emob-nordic — CHANGES_REQUESTED** [review]

File: subsys/caf/modules/Kconfig.click_detector (hunk 1 above)
Comment (MarekPieta) [reply subsys/caf/modules/Kconfig.click_detector:2]: done

File: subsys/caf/modules/Kconfig.click_detector (hunk 2 above)
Comment (MarekPieta) [reply subsys/caf/modules/Kconfig.click_detector:27]: done

File: include/caf/click_detector.rst (hunk 3 above)
Comment (MarekPieta) [reply include/caf/click_detector.rst:26]: There are no such options in this case. Only the module that submits button events is needed here (described in first sentence).

File: include/caf/click_detector.rst (hunk 4 above)
Comment (MarekPieta) [reply include/caf/click_detector.rst:50]: Added. Fixed.

File: include/caf/click_detector.rst (hunk 3 above)
Comment (emob-nordic) [reply include/caf/click_detector.rst:26]: There are. Summon @greg-fer.

```
To use the module, you must enable the following Kconfig options:

* :option:`CONFIG_CAF_CLICK_DETECTOR` - This option enables the click detector module.
* :option:`CONFIG_CAF_BUTTON_EVENTS` - This option enables the button events which are required for click detector module to work.

Comment (MarekPieta) [reply include/caf/click_detector.rst:26]: Ok, changed. Previously the Kconfig option that enabled the module was added as a final step of configuration.
Sorry for misunderstanding.

File: include/caf/click_detector.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+**********
+
+The module records the following click types:
+
+* :c:enumerator:`CLICK_SHORT` - Button pressed and released after short time.
+* :c:enumerator:`CLICK_NONE` - Button pressed and held for a period of time that is too long for :c:enumerator:`CLICK_SHORT`, but too short for :c:enumerator:`CLICK_LONG`.
+* :c:enumerator:`CLICK_LONG` - Button pressed and held for a long period of time.
+* :c:enumerator:`CLICK_DOUBLE` - Two sequences of the button press and release in a short time interval.
+
+The exact values of time intervals for click types are defined in the :file:`subsys/caf/modules/click_detector.c` file.
```
Comment (emob-nordic) [inline include/caf/click_detector.rst:22]: Just a reflection here. @padu shouldn't we make those make those time intervals configurable? In particular short and long press. If not per button, at least globally.

**Comment by emob-nordic** [comment]
I remind myself one thing.
There are tables in nrf_desktop that describes event relations for each nrf_desktop module (located in `event_propagation.rst` file).
In the current state of things (CAF buttons module) those tables relates directly to CAF modules.
However You left those relations as is.
The question is:
Do we change existing buttons module or adapt `click_detector` to follow existing style? @pdunaj  @greg-fer.

**Comment by pdunaj** [comment]
> I remind myself one thing.
> There are tables in nrf_desktop that describes event relations for each nrf_desktop module (located in `event_propagation.rst` file).
> In the current state of things (CAF buttons module) those tables relates directly to CAF modules.
> However You left those relations as is.
> The question is:
> Do we change existing buttons module or adapt `click_detector` to follow existing style? @pdunaj @greg-fer.

This is a difficult thing that we will have to review (I don't think this PR is a good place as we would need to smoothen things after moving the modules).
The tables mentioned were autogenerated by script checking the content of the source files. Tables belong to project (in this case these are part of nrf_desktop). It would be nice to keep the links to project related pages (e.g. buttons in nrf_desktop). The problem is not always such pages will be provided (as there will be no project specific module implementations). In the end we should make links to caf.

What we could do is split caf and project specific module implementations (it was done to some extend). When tables are generated script could check what CAF options are enabled and depending on it add tables required. It could be possible to merge the results by analyzing module names (so e.g. motion_sensor and motion_sim could go to a single table for one project /assuming motion_sensor will be generalized/).

I propose to do as it was done so far during the migration, that is fix links in tables by using references to caf. Later on we will talk about it and decide how to update the script so the tables will remain autogenerated.

**Review by greg-fer — APPROVED** [review]

**Review by emob-nordic — APPROVED** [review]

**Pull request merged**